Add some design considerations and reasoning (#4)
SovereignCloudStack/central-api: README.md

# SCS Central API

## Premise

By embracing existing open source solutions and bundling them, SCS is intended to
provide a viable alternative to widely adopted proprietary cloud offerings, including
Infrastructure-as-a-Service offerings, Kubernetes-as-a-Service offerings and other
X-as-a-Service offerings.

The choice to embrace existing technology has huge advantages over starting from
scratch.
By not reinventing wheels, a lot of effort can be saved. Existing communities can
be strengthened. The adoption of existing open standards is supported, reducing

> market fragmentation and increasing interoperability.

Reviewer: while this is imho a good intro for SCS in general I'm not 100% sure it's needed as an intro for the central api? I have no strong opinion on removing it though.

Author: I guess the "premise" section is not strictly required, but IMHO helps a lot clarifying scope and the current situation.

In particular, in relation to the central API, the first paragraph hints at...
- the scope of services within SCS, hence the expected scope of a central API
- the need to provide an UX on-par with hyperscalers, where central API's IMHO play an important role.

The second paragraph stresses the benefits of the chosen approach to build a cloud stack. Yes, technically, it's inconsequential here, but I wanted to start off on a positive note here, as the "challenge" section highlights the problems in much more detail.

Reviewer 2: i welcome that paragraph, to build on existing approaches has undoubtedly many advantages and it is a good starting point to assure that we are on the same page regarding this.

## Challenge

The challenge: Using popular open source components at cloud service providers
does not result in a consistent experience for their users, yet.

Each part of the stack is consistent within its own scope: E.g. The
[OpenStack Networking API](https://docs.openstack.org/api-ref/network/v2/) is sort of
consistent with the
[OpenStack Load Balancer API](https://docs.openstack.org/api-ref/load-balancer/v2/).

The OpenStack API's share API idioms like the used AuthN/AuthZ mechanisms. But these
are not applicable beyond OpenStack services.

> Entering general IAM, Keycloak has its own set of API endpoints and authentication

Reviewer: acronyms like IAM should imho be explained/expanded the first time they are used (or linked to a glossary).

Author: Fixed in fe98e9af38e1bad334ffb1d6ae12e4f316cb485d.

Reviewer 3: If possible, we should apply this to the other acronyms as well: 

* AuthN
* AuthZ
* CAPI

If you say "*are basically an abstraction over CAPI/ClusterAPI*", do you mean CAPI and ClusterAPI are two separate things? :wink:

Author: > If possible, we should apply this to the other acronyms as well:
> 
> * AuthN
> * AuthZ
> * CAPI

Did that in a66f1ab74a379aa028259957a95c2f2a653629c8. Missed the "acronyms **like**"  the first time around.

> If you say "_are basically an abstraction over CAPI/ClusterAPI_", do you mean CAPI and ClusterAPI are two separate things? 😉

No :D
Fixed that in above commit as well.

not seam feasible given tough trade-offs and limited resources.  
If the opportunity arises to partner with some other organization with a lot
of staff and resources, this option may be reevaluated, though.

</details>

## The chosen approach (for now)

Goal: **Provide a "semantically" consistent API modelling most cloud resources
that are in scope for SCS**.

### Kubernetes API

Instead of creating SCS-specific API idioms and building the implementation
from scratch, the Kubernetes API will be "reused". Essentially, the Kubernetes
API is just an opinionated REST API which has opinions on how a resource
is defined, how it looks like, how it is reconciled/handled, how AuthN/AuthZ
can be implemented. The Kubernetes ecosystem provides much tooling for working
with such (custom) resource definitions: For creating the definitions
themselves, building controllers, making them discoverable and deployable.

As such, Kubernetes is a great choice for building any sort of resource
management API - with some caveats regarding its deployment and the legacy
of starting off as container orchestration tooling.

### Crossplane tooling

> Crossplane even extends the Kubernetes API with "Composite Resource Definitions"

Reviewer: it might be good to link to the definition here for readers that are not familiar with this crossplane implementation detail?

I believe this link should maybe do the job? https://docs.crossplane.io/latest/concepts/compositions/ (maybe better link to a specific version in case the "latest" docs might change in the future).

Author: Did so in 96833561574624eb58b63bbd27a72159829ab34e.

(XRD) to make Kubernetes the base for platform engineering within organizations.

Secondly, it provides a API machinery to bring any cloud resource into Kubernetes
using backend-specific "providers" (roughly comparable with Terraform providers).
As such, Crossplane with its provider ecosystem actually already did most of
the heavy lifting for providing e.g. OpenStack resources inside of Kubernetes.

On top, the platform engineering concepts in Crossplane make building multi-tenancy
systems pretty straight-forward.

Alright. Crossplane takes care of exposing OpenStack resources and does some
fancy stuff regarding multi-tenancy. What about providing actual Kubernetes
**workload** clusters?

### Cluster stacks

> Cluster stacks are basically an abstraction over CAPI/ClusterAPI. As such, Cluster

Reviewer 2: Clusterstacks are mostly tooling that is relevant for the CSP not for the SCS-user. The interesting point for the SCS-User is that there will be(if we do it right) a documented, public repository/list of clusterclasses (packed in cluster-stacks). Those clusterclasses are of interest for the SCS-User because they define variables that shape the cluster (flavor, dns-settings). Also the choice of the clusterstack/clusterclass will have an impact on the behaviour of the cluster.

Author: Ah, alright. Thanks for clarifying. Also had a look at the [cluster-stack-operator docs](https://github.com/SovereignCloudStack/cluster-stack-operator/blob/main/docs/architecture/user-flow.md) now, improving my understanding of cluster stacks. Will change the wording here.

Author: Wording is changed.

Author: @mxmxchere This can be resolved from your POV?

For example:

Provider "A" offers to hide Kubernetes API server endpoints from the public
internet, utilizing some sort of bastion host. Provider "B" instead implements
IP based firewall blocking on the public endpoints. Provider "C" does neither.  
Should the API follow either provider "A" or "B"? Should both approaches be
implemented, but as optional features? If any of these approaches is defined
to be a mandatory feature to support, provider "C" cannot be compliant.

Any choice brings significant disadvantages:

- If such features are included as features that are mandatory to support,
  some providers may have difficulties adopting the API.
- If such features are included as optional features, the ability to migrate
  from one provider to another suffers significantly. Without this ability,
  users also may opt to use provider-specific API's, instead.
- If such features are excluded, the API becomes overall less useful for the
  users who may opt to use more powerful provider-specific API's, instead.

In sum: Going this route would be technically the best thing to do, yet does
not seem feasible given tough trade-offs and limited resources.  

> If the opportunity arises to partner with some other organization with a lot

Reviewer 4: Reading the previous paragraph, I don't understand this conclusion. Do you think that it is better to have stronger standards that are better for users but more difficult for providers, or do you think the opposite is better?

Author: As you already said: Both (polar) options are difficult/impracticable/useless for either provider or user. Everything in-between is not about finding the perfect solution for everyone, but "just the right" balance that is practicable and valuable for both providers and users.

Finding this balance for all to-be-covered features would probably cost much more time than there is available.

Reviewer 4: what do you mean by "all to-be-covered features"?

Author: As the intro says: "Infrastructure-as-a-Service offerings, Kubernetes-as-a-Service offerings and other X-as-a-Service offerings."

Meaning that for OpenStack/IaaS alone, one would need to decide on "just the right" balance for each of the [97 resources](https://marketplace.upbound.io/providers/crossplane-contrib/provider-openstack/v0.3.0) that the crossplane openstack provider exposes. Building a new abstraction that should hide any OpenStack implementation details and yet be useful. Possibly merging some objects, possibly skipping features; Definitely causing much work.

Reviewer 4: now you sound as if you talk about both ways with trade-offs. In the document you say "going this route", which seems to be only one

Author: "this route" is meant to refer to the strategy to "Abstract away all the things!" (topic of this part), hence trying to find the optimal tradeoffs for everything.

Author: Attempted to make this clearer in the document in dce79ed134e157fc27aff172a256f5328eaf1e3e.

Does this address our misunderstanding appropriately?

Reviewer 4: definitely. I am still missing a bit the relation to the chosen approach though. Given the fact that we will not be able to balance the two things easily (this is what you wrote), what exactly is the approach that Central API chooses?

Author: Spelled that out more explicitly in 63547eb4a3c73876f3aeb61de065152d216da0b7.

Does this answer the question?

Reviewer 4: it does!

of starting off as container orchestration tooling.

### Crossplane tooling

Crossplane even extends the Kubernetes API with
"[Compositions](https://docs.crossplane.io/v1.14/concepts/compositions/)" and
"[Composite Resource Definitions](https://docs.crossplane.io/v1.14/concepts/composite-resource-definitions/)"
(XRD) to make Kubernetes the base for platform engineering within organizations.

Secondly, it provides an API machinery to bring any cloud resource into Kubernetes
using backend-specific "providers" (roughly comparable with Terraform providers).
As such, Crossplane with its provider ecosystem actually already did most of
the heavy lifting for providing e.g. OpenStack resources inside of Kubernetes.

On top, the platform engineering concepts in Crossplane make building multi-tenancy
systems pretty straight-forward, even for
[single clusters](https://docs.crossplane.io/knowledge-base/guides/multi-tenant/#single-cluster-multi-tenancy).

Alright. Crossplane takes care of exposing OpenStack resources and does some
fancy stuff regarding multi-tenancy. What about providing actual Kubernetes
**workload** clusters?

### Cluster stacks

> Cluster stacks are basically an abstraction over Cluster API. As such, Cluster

Reviewer 4: I agree with @mxmxchere that this part has to be written better. I would in general and very high-level say that Cluster Stacks make it easier to work with Cluster API with various features, like tested upgrades, creating n clusters from the same template, etc.

Author: As in @mxmxchere's review [above](https://github.com/SovereignCloudStack/central-api/pull/4#discussion_r1484147650), thanks for clarifying. Will change that wording.

Author: Wording is changed.

(XRD) to make Kubernetes the base for platform engineering within organizations.

Secondly, it provides an API machinery to bring any cloud resource into Kubernetes
using backend-specific "providers" (roughly comparable with Terraform providers).
As such, Crossplane with its provider ecosystem actually already did most of
the heavy lifting for providing e.g. OpenStack or Keycloak resources inside of Kubernetes.

On top, the platform engineering concepts in Crossplane make building multi-tenancy
systems pretty straight-forward, even for
[single clusters](https://docs.crossplane.io/knowledge-base/guides/multi-tenant/#single-cluster-multi-tenancy).

Alright. Crossplane takes care of exposing OpenStack resources and does some
fancy stuff regarding multi-tenancy. What about providing actual Kubernetes
**workload** clusters?

### Cluster stacks / Cluster API

[Cluster stacks](https://github.com/SovereignCloudStack/cluster-stacks) do
[not replace the use of Cluster API](https://github.com/SovereignCloudStack/cluster-stack-operator/blob/adb648ceaebddca04a015fbea0319110ca99a5cc/docs/architecture/user-flow.md#recap---how-do-cluster-api-and-cluster-stacks-work-together).

> Instead, they are complementing Cluster API by providing `ClusterClasses`.

Reviewer 4: ... as well as node images (if required) and cluster addons in workload clusters!

Author: For the purposes of this document, I saw node images and cluster addons more like something that a `ClusterClass` provides/embeds.

But, sure, mentioning it does not hurt. Created 414b773.

Reviewer 4: no clusterclass is definitely not related to the cluster addons. ok!

> It is still to be determined how to bring multi-tenancy concepts from Crossplane

Reviewer 4: what do you mean by that? Multi-tenancy here should work with different namespaces

Author: Crossplane enables multi-tenancy in distinct ways, for e. g. [Single Cluster Multi-Tenancy](https://docs.crossplane.io/knowledge-base/guides/multi-tenant/#single-cluster-multi-tenancy).
I'm not sure yet, whether integrating CAPI/ClusterStacks with Crossplane's approaches makes sense.

Also, I did not operate a multi-tenant CAPI environment yet, so I'm not sure if namespace isolation is enough for providing multi-tenant CAPI.

Reviewer 4: seems like Crossplane also has namespace-based isolation, no? just quickly went through the docs you referenced

Author: Yes, but in a special manner. Namespaced resources that can be created by lower-privileged users are kind of like "claims" (like a PVC) that result in the creation of cluster-scoped resources (like a PV).

Also see the contents that I moved into `poc-setup.md` in this PR.

Reviewer 4: mmh anyway, since this is not completely clear (cluster api and/vs crossplane) and that's what you wrote in there, I guess we can just resolve it

# SCS Central API

## Premise

> By embracing existing open source solutions and bundling them, SCS is intended to
> provide a viable alternative to widely adopted proprietary cloud offerings, including

Reviewer 5: ```suggestion
By embracing existing open source solutions and bundling them, SCS provides a viable alternative to widely adopted proprietary cloud offerings, including
```

Infrastructure-as-a-Service offerings, Kubernetes-as-a-Service offerings and other
X-as-a-Service offerings.

The choice to embrace existing technology has huge advantages over starting from
scratch.

> By not reinventing wheels, a lot of effort can be saved. Existing communities can
> be strengthened. The adoption of existing open standards is supported, reducing
> market fragmentation and increasing interoperability.

Reviewer 5: ```suggestion
By not reinventing wheels, a lot of effort is saved and existing communities are strengthened. The adoption of existing open standards is supported, reducing
market fragmentation and increasing interoperability.
```

OpenStack services.

Entering general IAM (Identity and Access Management), Keycloak has its own set of
API endpoints and authentication flows.  
Entering Kubernetes, CAPI ([Kubernetes Cluster API](https://cluster-api.sigs.k8s.io/))
uses the Kubernetes API with its own authentication configuration, RBAC (Role Based
Access Control) and opinionated resource management idioms.

So, without a central API harmonizing at least the semantics of AuthN/AuthZ and
resource management, users are left with a bunch of semantically incompatible API's.
If resources in different API's are somehow interconnected, the users have to take
care of bridging these differences themselves.

Providing a consistent API across many different offerings with sort of consistent
API idioms is something that primarily the big proprietary cloud providers manage to
do. And while that serves users well in that regard, it also serves as an effective
vendor lock-in feature.

## Alternatives and other potential solutions

<details><summary>Self-Service solution</summary>

### Self-Service solution

Users that want to avoid such vendor lock-in as well as want to avoid spending much

> time bridging technologies manually, the best bet would probably be to setup
> e.g. Terraform (or OpenTofu)
> with a number of specialized providers to bring all their interdependent resources into

Reviewer 5: ```suggestion
time bridging technologies manually, the best bet would probably be to setup infrastructure-as-code (IaC) tooling (such as OpenTofu, Terraform and alike)
with a number of specialized providers to bring all their interdependent resources into
```

a single place keeping track of relationships between resources across multiple API's.

> Caveat: Terraform/OpenTofu/... gets admin access, while RBAC for human access is still

Reviewer 5: ```suggestion
Caveat: Infrastructure-as-code tooling gets admin access, while RBAC for human access is still
```

defined using consistent patterns and terminology, never leaking implementation details.

- `OpenStack Compute Instance`? Very OpenStack specific, creating lock-in to OpenStack API's
- `SCS Instance`? Perfection, right?

Imagine CLI access like:

```bash
scs create subnet --group foo
scs create k8s --group foo my-k8s-01
```

Imagine using a Terraform provider like:

```hcl
resource "scs_group" "mygroup" {
  name = "mygroup"
}
resource "scs_subnet" "mysubnet" {
  group = scs_group.name
  name  = "mysubnet"
  # ...
}
resource "scs_kubernetes" "mykubernetes" {
  group  = scs_group.name
  name   = "mykubernetes"
  subnet = scs_subnet.mysubnet.id
  # ...
}
```

Imagine a Crossplane provider (or DIY similar Kubernetes controller framework) like:

```yaml
apiVersion: networking.scs.community/v1
kind: Subnet
metadata:
  name: mysubnet
spec:
  # ...
---
apiVersion: kubernetes.scs.community/v1
kind: Kubernetes
metadata:
  name: mykubernetes
spec:
  forProvider:
    subnetRef:
      name: mysubnet
  # ...
```

This is obviously desirable from a user's perspective.
However, unfortunately, it is also much more work than the SCS project can

> realistically build and maintain in the short/medium term.

Reviewer 5: ```suggestion
realistically build and maintain in the short or medium term.
```
